src/conn/conn_api.c

```diff
@@ -1436,11 +1440,19 @@ __conn_write_config(
 		WT_ERR_NOTFOUND_OK(ret);
 	}
 
-err:	WT_TRET(fclose(fp));
+	/* Flush to disk and close our handle. */
+	WT_ERR(__wt_fp_close(session, &fp));
 
-	/* Don't leave a damaged file in place. */
-	if (ret != 0)
-		(void)__wt_remove(session, filename);
+	/*
+	 * Rename the file into place: that's an atomic operation so we can be
+	 * confident we never see a partial file.
+	 */
+	WT_ERR(__wt_rename(session, WT_BASECONFIG_SET, WT_BASECONFIG));
+
+err:	WT_TRET(__wt_fp_close(session, &fp));
+
+	/* Discard any damaged temporary file, not required but cleaner. */
+	WT_TRET(__wt_remove_if_exists(session, WT_BASECONFIG_SET));
```

Comment:
> On Mar 20, 2015, at 17:23, Keith Bostic <notifications@github.com> wrote:
> 
> In src/conn/conn_api.c <https://github.com/wiredtiger/wiredtiger/pull/1785#discussion_r26877654>:
> 
> >  
> > -	/* Don't leave a damaged file in place. */
> > -	if (ret != 0)
> > -		(void)__wt_remove(session, filename);
> > +	/*
> > +	 * Rename the file into place: that's an atomic operation so we can be
> > +	 * confident we never see a partial file.
> > +	 */
> > +	WT_ERR(__wt_rename(session, WT_BASECONFIG_SET, WT_BASECONFIG));
> > +
> > +err:	WT_TRET(__wt_fp_close(session, &fp));
> > +
> > +	/* Discard any damaged temporary file, not required but cleaner. */
> > +	WT_TRET(__wt_remove_if_exists(session, WT_BASECONFIG_SET));
> I think fflush implies fsync, from http://pubs.opengroup.org/onlinepubs/009695399/functions/fflush.html <http://pubs.opengroup.org/onlinepubs/009695399/functions/fflush.html>:
> 
> The functionality described on this reference page is aligned with the ISO C standard. Any conflict between the requirements described here and the ISO C standard is unintentional. This volume of IEEE Std 1003.1-2001 defers to the ISO C standard.
> 
> If stream points to an output stream or an update stream in which the most recent operation was not input, fflush() shall cause any unwritten data for that stream to be written to the file, and the st_ctime and st_mtime fields of the underlying file shall be marked for update.
> 
> Is there something I'm missing?
> 
Keith, I see fflush man page for linux says something different, from: http://linux.die.net/man/3/fflush:
Notes

Note that fflush() only flushes the user-space buffers provided by the C library. To ensure that the data is physically stored on disk the kernel buffers must be flushed too, for example, with sync <http://linux.die.net/man/2/sync>(2) or fsync <http://linux.die.net/man/2/fsync>(2).

Perhsps ‘written to the file’ in the OpenGroup spec just means write(2), not physical write.

- Don

--

Don Anderson
C/C++/Java, systems consultant

voice: 617-306-2057
email: dda@ddanderson.com <mailto:dda@ddanderson.com>
www: http://www.ddanderson.com <http://www.ddanderson.com/>
blog: http://libdb.wordpress.com <http://libdb.wordpress.com/>